src/utils/changes.ts

```diff
@@ -175,3 +181,152 @@ export function prependChangeset(
 
   return markdown.slice(0, startIdx) + newSection + markdown.slice(startIdx);
 }
+
+interface Commit {
+  hash: string;
+  title: string;
+  hasPRinTitle: boolean;
+  pr: string | null;
+  milestone: string | null;
+}
+
+interface Milestone {
+  title?: string;
+  description?: string;
+  state?: string;
+  prs: string[];
+}
+
+// This is set to 8 since GitHub and GitLab prefer that over the default 7 to
+// avoid collisions.
+const SHORT_SHA_LENGTH = 8;
+function formatCommit(commit: Commit): string {
+  let text = `- ${commit.title}`;
+  if (!commit.hasPRinTitle) {
+    const link = commit.pr
+      ? `#${commit.pr}`
+      : commit.hash.slice(0, SHORT_SHA_LENGTH);
+    text = `${text} (${link})`;
+  }
+  return text;
+}
+
+export async function generateChangesetFromGit(
+  git: SimpleGit,
+  rev: string
+): Promise<string> {
+  const { repo, owner } = await getGlobalGithubConfig();
+  const gitCommits = await getChangesSince(git, rev);
+  const commitsQuery = gitCommits
+    .map(({ hash }) => `C${hash}: object(oid: "${hash}") {...PRFragment}`)
+    .join('\n');
+
+  const github = getGithubClient();
+  const { repository: githubCommits } = await github.graphql(`{
+    repository(name: "${repo}", owner: "${owner}") {
+      ${commitsQuery}
+    }
+  }
+
+  fragment PRFragment on Commit {
+    associatedPullRequests(first: 1) {
+      nodes {
+        number
+        milestone {
+          number
+        }
+      }
+    }
+  }`);
+
+  const milestones: {
+    [number: string]: Milestone;
+  } = {};
+  const commits: {
+    [hash: string]: Commit;
+  } = {};
+  const leftovers: Commit[] = [];
+  const missing: Commit[] = [];
+  for (const gitCommit of gitCommits) {
+    const hash = gitCommit.hash;
+
+    const commit = (commits[hash] = {
+      hash: hash,
+      title: gitCommit.message,
+      pr: gitCommit.pr,
+      hasPRinTitle: Boolean(gitCommit.pr),
+      milestone: null,
+    });
+    const commitInfo = githubCommits[`C${hash}`];
+    if (commitInfo) {
+      const pr = commitInfo.associatedPullRequests.nodes[0];
+      if (pr) {
+        commit.pr = pr.number;
+        commit.milestone = pr.milestone?.number;
+      }
```

Comment:
IMO, moving this before `commit` is created makes it easier to read (get the data and create the object from it, instead of creating it with some data and assigning them afterward). Same with the milestone logic below.

Reply:
Not sure if that's actually possible but I'll try.